File: app/src/main/java/com/example/client/screens/reset/present/PasswordResetPresent.java
package com.example.client.screens.reset.present;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.widget.TextView;

import com.example.client.api.ApiClient;
import com.example.client.api.service.UserService;
import com.example.client.app.Constants;
import com.example.client.models.message.MessageModel;
import com.example.client.screens.reset.activity.IPasswordResetView;

import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class PasswordResetPresent implements IPasswordResetPresent{
    private IPasswordResetView pView;
    public PasswordResetPresent(IPasswordResetView pView){
        this.pView = pView;
    }

    @Override
    public void vertification(String email, String code) {
        pView.showVertifiLoading();
        UserService service = ApiClient.getInstance().create(UserService.class);
        service.vertification(email,code).enqueue(new Callback<MessageModel>() {
            @Override
            public void onResponse(Call<MessageModel> call, Response<MessageModel> response) {
                pView.showViewPassword(response.body());
                pView.hideVertifiLoading();
            }

            @Override
            public void onFailure(Call<MessageModel> call, Throwable t) {
                pView.showViewPassword(new MessageModel(false,1001,null));
                pView.hideVertifiLoading();
            }
        });

    }

    @Override
    public void onResetPassword(String email, String password) {
        pView.showResetLoading();
        UserService service = ApiClient.getInstance().create(UserService.class);
        service.resetPassword(email,password).enqueue(new Callback<MessageModel>() {
            @Override
            public void onResponse(Call<MessageModel> call, Response<MessageModel> response) {
                pView.resetPassword(response.body());
                pView.hideResetLoading();
            }

            @Override
            public void onFailure(Call<MessageModel> call, Throwable t) {
                pView.resetPassword(new MessageModel(false,1001,null));
                pView.hideResetLoading();
            }
        });
    }

    @Override
    public void onSendEmail(String email) {
        pView.showSendEmailLoading();
        UserService service = ApiClient.getInstance().create(UserService.class);
        service.sendEmail(email, Constants.RequestType.RESET_PASSWORD).enqueue(new Callback<MessageModel>() {
            @Override
            public void onResponse(Call<MessageModel> call, Response<MessageModel> response) {
                pView.sendEmail(response.body());
                pView.hideSendEmailLoading();
            }

            @Override
            public void onFailure(Call<MessageModel> call, Throwable t) {
                pView.sendEmail(new MessageModel(false,1001,null));
                pView.hideSendEmailLoading();
            }
        });
    }

    @SuppressLint("CheckResult")
    @Override
    public void onCountDownTimer(TextView tv, Integer time) {
        Observable.interval(1, TimeUnit.SECONDS)
                .take(time)
                .map(v -> v + 1 )
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(value->{
                    value = time - value;
                    int minutes = (int) ((value % 3600) / 60);
                    int seconds = (int) (value % 60);
                    String timeString = String.format("%02d:%02d", minutes, seconds);
                    tv.setText("Gửi mail xác thực (" + timeString + ")");
                    tv.setTextColor(Color.GRAY);
                    tv.setEnabled(false);
                },Throwable::printStackTrace,() -> {
                    tv.setText("Gửi mail xác thực");
                    tv.setTextColor(Color.BLUE);
                    tv.setEnabled(true);
                });
    }
}

File: app/src/main/java/com/example/client/screens/reset/present/PasswordResetPresent.kt
package com.example.client.screens.reset.present

import android.graphics.Color
import android.widget.TextView
import com.example.client.R
import com.example.client.app.Constants
import com.example.client.app.RxBus
import com.example.client.base.BasePresenterMVP
import com.example.client.models.event.Event
import com.example.client.screens.reset.activity.IPasswordResetView
import com.example.client.usecase.ProfileUseCase
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import java.text.MessageFormat
import java.util.*
import java.util.concurrent.TimeUnit

class PasswordResetPresent(mView: IPasswordResetView) : BasePresenterMVP<IPasswordResetView>(mView), IPasswordResetPresent {
    private val profileUseCase by lazy { ProfileUseCase.newInstance() }

    override fun sendEmail(email: String) {
        mView?.showLoading()
        subscribe(profileUseCase.sendEmail(email, "000", Constants.RequestType.RESET_PASSWORD), {
            mView?.run {
                hideLoading()
                if (it.is_error) {
                    if (it.code == 1010) sendMailSuccess()
                    showErrorMessage(getErrorMessage(it.code))
                    return@subscribe
                }
                sendMailSuccess()
            }
        }, {
            it.printStackTrace()
            mView?.run {
                hideLoading()
                showErrorMessage(getErrorMessage(1001))
            }
        })
    }

    override fun verification(email: String, code: String) {
        mView?.showLoading()
        subscribe(profileUseCase.verification(email, code), {
            mView?.run {
                hideLoading()
                if (it.is_error) {
                    showErrorMessage(getErrorMessage(it.code))
                    return@subscribe
                }
                verificationSuccess()
            }
        }, {
            it.printStackTrace()
            mView?.run {
                hideLoading()
                showErrorMessage(getErrorMessage(1001))
            }
        })
    }

    override fun resetPass(email: String, password: String) {
        mView?.showLoading()
        subscribe(profileUseCase.resetPass(email, password), {
            mView?.run {
                hideLoading()
                if (it.is_error) {
                    showErrorMessage(getErrorMessage(it.code))
                    return@subscribe
                }
                resetPassSuccess()
                RxBus.newInstance().onNext(Event(Constants.EventKey.RESET_SUCCESS))
            }
        }, {
            it.printStackTrace()
            mView?.run {
                hideLoading()
                showErrorMessage(getErrorMessage(1001))
            }
        })
    }

    override fun countDownTimer(tv: TextView, time: Int) {
        add(Observable.interval(1, TimeUnit.SECONDS)
                .take(time.toLong())
                .map { v: Long -> v+1 }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    val value = time - it
                    val minutes = (value % 3600 / 60).toInt()
                    val seconds = (value % 60).toInt()
                    val timeString = String.format(Locale.getDefault(), "%02d:%02d", minutes, seconds)
                    tv.text = MessageFormat.format("Gửi mail xác thực ({0})", timeString)
                    tv.setTextColor(Color.GRAY)
                    tv.isEnabled = false
                },{
                    it.printStackTrace()
                    tv.setText(R.string.reset_text_send_email)
                    tv.setTextColor(Color.BLUE)
                    tv.isEnabled = true
                }))
    }

}